receiver/githubreceiver/logs.go

```diff
@@ -0,0 +1,315 @@
+
+		// timestamps
+		logRecord.SetObservedTimestamp(observedTime)
+		timestamp := time.UnixMilli(logEvent.Timestamp)
+		logRecord.SetTimestamp(pcommon.NewTimestampFromTime(timestamp))
+
+		// body
+		logEventBytes, err := json.Marshal(logEvent)
+		if err != nil {
+			r.logger.Error("unable to marshal logEvent", zap.Error(err))
+		} else {
+			logRecord.Body().SetStr(string(logEventBytes))
+		}
+		// add attributes
+		r.addAttributes(logRecord, logEvent)
+	}
+	return logs
+}
+
+// helper function to add attributes to logRecord
+func (r *gitHubLogsReceiver) addAttributes(logRecord plog.LogRecord, logEvent gitHubEnterpriseLog) {
+	// add attributes
+	attrs := logRecord.Attributes()
+	attrs.PutInt("@timestamp", logEvent.Timestamp)
+	attrs.PutStr("_document_id", logEvent.DocumentID)
+	attrs.PutStr("action", logEvent.Action)
+	attrs.PutStr("actor", logEvent.Actor)
+	attrs.PutInt("actor_id", logEvent.ActorID)
+	attrs.PutBool("actor_is_bot", logEvent.ActorIsBot)
+	attrs.PutStr("actor_location", logEvent.ActorLocation.CountryCode)
+	attrs.PutStr("business", logEvent.Business)
+	attrs.PutInt("business_id", logEvent.BusinessID)
+	attrs.PutInt("created_at", logEvent.CreatedAt)
+	attrs.PutStr("operation_type", logEvent.OperationType)
+	attrs.PutStr("user_agent", logEvent.UserAgent)
+
+	// map of optional attributes
+	optionalAttrs := map[string]interface{}{
+		"name":                    logEvent.Name,
+		"org":                     logEvent.Org,
+		"org_id":                  logEvent.OrgID,
+		"organization_upgrade":    logEvent.OrganizationUpgrade,
+		"permission":              logEvent.Permission,
+		"user":                    logEvent.User,
+		"user_id":                 logEvent.UserID,
+		"owner_type":              logEvent.OwnerType,
+		"audit_log_stream_sink":   logEvent.AuditLogStreamSink,
+		"audit_log_stream_result": logEvent.AuditLogStreamResult,
+	}
+
+	// add optional attributes if they are present
+	for key, value := range optionalAttrs {
+		switch v := value.(type) {
+		case string:
+			if v != "" {
+				attrs.PutStr(key, v)
+			}
+		case int64:
+			if v != 0 {
+				attrs.PutInt(key, v)
+			}
+		case bool:
+			if v {
+				attrs.PutBool(key, v)
+			}
+		}
+	}
+
+}
+func (r *gitHubLogsReceiver) setNextLink(res *http.Response, page int) int {
+	for _, link := range res.Header["Link"] {
+		// Split the link into URL and parameters
+		parts := strings.Split(strings.TrimSpace(link), ";")
+		if len(parts) < 2 {
+			continue
+		}
+		
+		// Check if the "rel" parameter is "next"
+		if strings.TrimSpace(parts[1]) == `rel="next"` {
+			// increment page
+			page ++
+			// Extract and return the URL
+			r.nextURL = strings.Trim(parts[0], "<>")
+			
+			return page
+		}
+	}
+	r.logger.Error("unable to get next link")
+	return page
+}
+
+func millisecondsToTime(ms int64) time.Time {
+    seconds := ms / 1000
+    nanoseconds := (ms % 1000) * 1000000
+    return time.Unix(seconds, nanoseconds)
+}
```

Comment:
This may be a bit simpler way to convert milliseconds to Time, I was just curious and did some googling/chatGPTing. Not saying you have to change it though just found it interesting.

```
func millisecondsToTime(ms int64) time.Time {
    return time.Unix(0, ms*int64(time.Millisecond))
}
```